scikeras/wrappers.py

```diff
@@ -627,16 +628,22 @@ def _fit(self, X, y, sample_weight=None, warm_start=False):
             self._random_state = self.random_state
 
         # Data checks
-        if warm_start and not hasattr(self, "n_features_in_"):
-            # Warm start requested but not fitted yet
-            reset = True
-        elif warm_start:
-            # Already fitted and warm start requested
-            reset = False
+        should_init = True
+        if (self.warm_start or warm_start) and self._initialized():
+            should_init = False
+        if should_init:
+            X, y, meta = self._initialize(X, y)
         else:
-            # No warm start requested
-            reset = True
-        X, y = self._validate_data(X=X, y=y, reset=reset)
+            X, meta = self.preprocess_X(X)
+            y, _ = self.preprocess_y(y)
+        self._check_output_model_compatibility(y)
+
+        if meta["n_features_in_"] != self.n_features_in_:
+            raise ValueError(
+                f"`X` has {meta['n_features_in_']} features, but this "
+                f"{self.__name__} is expecting {self.n_features_in_} "
+                f"features as input."
+            )
```

Comment:
This check is kind of lonely. Shouldn't we be checking all of `meta`'s keys (where it makes sense to do so)?

Reply:
I moved this check to `preprocess_X`, where we were already checking `meta["X_dtype_"]`.